src/main/java/com/squareup/javapoet/CodeWriter.java

```diff
@@ -272,106 +272,80 @@ private void emitLiteral(Object o) throws IOException {
    * names visible due to inheritance.
    */
   String lookupName(ClassName className) {
-    // Different package than current? Just look for an import.
-    if (!className.packageName().equals(packageName)) {
-      if (conflictsWithLocalName(className)) {
-        return className.toString(); // A local name conflicts? Use the fully-qualified name.
-      }
-
-      String importedName = importedTypes.get(className);
-      if (importedName != null) {
-        if (!javadoc) importableTypes.add(className);
-        referencedNames.add(importedName);
-        return importedName;
-      }
-
-      // If the target class wasn't imported, perhaps its enclosing class was. Try that.
-      ClassName enclosingClassName = className.enclosingClassName();
-      if (enclosingClassName != null) {
-        return lookupName(enclosingClassName) + "." + className.simpleName();
+    // Find the shortest suffix of className that resolves to className. This uses both local type
+    // names (so `Entry` in `Map` refers to `Map.Entry`). Also uses imports.
+    boolean nameResolved = false;
+    for (ClassName c = className; c != null; c = c.enclosingClassName()) {
+      ClassName resolved = resolve(c.simpleName());
+      nameResolved = resolved != null;
+
+      if (Objects.equals(resolved, c)) {
+        int suffixOffset = c.simpleNames().size() - 1;
+        return join(".", className.simpleNames().subList(
+            suffixOffset, className.simpleNames().size()));
       }
+    }
 
-      // Fall back to the fully-qualified name. Mark the type as importable for a future pass.
-      if (!javadoc) importableTypes.add(className);
-      return className.canonicalName;
+    // If the class is in the same package, we're done.
+    if (Objects.equals(packageName, className.packageName())) {
+      referencedNames.add(className.topLevelClassName().simpleName());
+      return join(".", className.simpleNames());
     }
 
-    // Look for the longest non-conflicting common prefix, which we can omit.
-    List<String> classNames = className.simpleNames();
-    int prefixLength = nonConflictingCommonPrefixLength(classNames);
-    if (prefixLength == classNames.size()) {
-      return className.simpleName(); // Special case: a class referring to itself!
+    // We'll have to use the fully-qualified name. Mark the type as importable for a future pass.
```

Comment:
We actually execute `CodeWriter` 2x per file. On the first pass we don't import anything, but we do remember what we encountered. On the second pass, great success.